src/test/groovy/unit/org/pih/warehouse/api/CycleCountApiControllerSpec.groovy

```diff
@@ -0,0 +1,281 @@
+            mockedCandidates.each { csvPrinter.printRecord(it.hashCode()) }
+            csvPrinter.flush()
+            return csvPrinter
+        }
+        controller.params.facilityId = FACILITY_ID
+        cycleCountService.getCandidates(filterParams, FACILITY_ID) >> mockedCandidates
+
+        when:
+        controller.getCandidates(filterParams)
+
+        then:
+        response.contentType.contains(CSV_FORMAT)
+        response.getHeader(HttpHeaders.CONTENT_DISPOSITION).startsWith("attachment; filename=\"CycleCountReport-")
+        response.status == HttpStatus.SC_OK
+        response.text.contains("ID")
+        response.text.contains(cycleCountCandidate1.hashCode() as String)
+        response.text.contains(cycleCountCandidate2.hashCode() as String)
+    }
+
+    void 'getPendingCycleCountRequests should return valid JSON'() {
+        given:
+        Query mockQuery = Stub(Query) {
+            list() >> [Mock(PendingCycleCountRequest),
+                       Mock(PendingCycleCountRequest)]
+        }
+        List<PendingCycleCountRequest> mockedPendingRequests = Spy(PagedResultList, constructorArgs: [mockQuery]) {
+            getTotalCount() >> mockQuery.list().size()
+        }
+        CycleCountCandidateFilterCommand cycleCountCandidateFilterCommand = Mock(CycleCountCandidateFilterCommand)
+        controller.params.facilityId = FACILITY_ID
+        cycleCountService.getPendingCycleCountRequests(cycleCountCandidateFilterCommand, FACILITY_ID) >> mockedPendingRequests
+
+        when:
+        controller.getPendingCycleCountRequests(cycleCountCandidateFilterCommand)
+
+        then:
+        response.contentType.contains(JSON_FORMAT)
+        response.status == HttpStatus.SC_OK
+        response.json.data.size() == 2
+        response.json.totalCount == 2
+    }
+
+    void 'createRequests should fail when batch has duplicated product'() {
+        given: "the following commands"
+        CycleCountRequestCommand cycleCountRequestCommand = new CycleCountRequestCommand(
+                product: Spy(Product) {
+                    getProductCode() >> PRODUCT_CODE
+                },
+                blindCount: true,
+        )
+        CycleCountRequestBatchCommand cycleCountRequestBatchCommand = new CycleCountRequestBatchCommand(requests: [cycleCountRequestCommand])
+
+        and: "the following facility id is passed as a param"
+        controller.params.facilityId = FACILITY_ID
+
+        and: "the following GORM methods are mocked"
+        CycleCountRequest.metaClass.static.findByProductAndFacilityAndStatusNotInList = {
+            Product product, Location location, List<CycleCountRequestStatus> statuses ->
+            return Mock(CycleCountRequest)
+        }
+        Location.metaClass.static.findById = {
+            String id -> Mock(Location)
+        }
+
+        when: "the command is invalid"
+        cycleCountRequestBatchCommand.validate()
```

Comment:
I don't think asserts like this register under a "when" block. Also shouldn't it be false since it fails validation? You'd want:
```
...

then:
assert !cycleCountRequestBatchCommand.validate()

when:
controller.createRequests(cycleCountRequestBatchCommand)

then:
thrown(ValidationException)
```
(I like to explicitly declare my asserts with "assert" in tests.)

Reply:
I didn't treat this as an assertion - after calling the `validate`, the errors are assigned to `CycleCountRequestCommand`, so that the errors are thrown in `createRequests`